File: lap/week7/01/circle.cpp
// This file contains the code for the routines in the text.
// It is not complete and will not compile.
#include "SortedType.h"
#include <iostream>

template<class ItemType>
void FindItem(NodeType<ItemType>* listData, 
              ItemType item,
              NodeType<ItemType>*& location, 
              NodeType<ItemType>*& predLoc,
              bool& found)
// Assumption: ItemType is a type for which the operators ''<'' and
//      "==" are defined as either an appropriate built-in type or a
//      class that overloads these operations.
// Pre: List is not empty.
// Post:If there is an element someItem whose key matches item's
//      key, then found = true; otherwise, found = false.
//      If found, location contains the address of someItem and
//      predLoc contains the address of someItem's predecessor;
//      otherwise, location contains the address of item's logical
//      successor and predLoc contains the address of item's
//      logical predecessor.
{
  bool moreToSearch = true;

  location = listData->next;
  predLoc = listData;
  found = false;

  while (moreToSearch && !found)
  {
    if (item < location->info)
      moreToSearch = false;
    else if (item == location->info)
      found = true;
    else
    {
      predLoc = location;
      location = location->next;
      moreToSearch = (location != listData->next);
    }
  }
}

template<class ItemType>
void SortedType<ItemType>::InsertItem(ItemType item)
{
  NodeType<ItemType>* newNode;
  NodeType<ItemType>* predLoc;
  NodeType<ItemType>* location;
  bool found;
  
  newNode = new NodeType<ItemType>;
  newNode->info = item;
  if (listData != NULL)
  {
    FindItem(listData, item, location, predLoc, found);
    newNode->next = predLoc->next;
    predLoc->next = newNode;
    // If this is last node in list, reassign listData.
    if (listData->info < item)
      listData = newNode;
  }
  else // Inserting into an empty list.
  {
    listData = newNode;
    newNode->next = newNode;
  }
  length++;
}

template<class ItemType>
void SortedType<ItemType>::DeleteItem(ItemType item)
{
  NodeType<ItemType>* location;
  NodeType<ItemType>* predLoc;
  bool found;
  
  FindItem(listData, item, location, predLoc, found);
  if (predLoc == location) // Only node in list?
    listData = NULL;
  else
  {
    predLoc->next = location->next;
    if (location == listData) // Deleting last node in list?
      listData = predLoc;
  }
  delete location;
  length--;
}

template<class ItemType>
void recursive_print(NodeType<ItemType> *node, NodeType<ItemType> *head)
{
  if (node->next != head)
    recursive_print(node->next, head);
  std::cout << node->info << "\n";
}

template<class ItemType>
void SortedType<ItemType>::PrintReverse()
{
  if (this->listData == NULL)
    return ;
  recursive_print(this->listData->next, this->listData->next);
}

File: pp/PP07/circle.py
class NodeType:
    """ Node Type """
    def __init__(self, item):
        self.info = item
        self.next = None

class CircularLL:
    def __init__(self):
        self.listData = None
        self.length = 0
        self.currentPos = None

    def is_full(self):
        try:
            location = NodeType("test")
            return False
        except:
            return True

    def length_is(self):
        return self.length

    def make_empty(self):
        while self.listData != None:
            tempPtr = self.listData.next
            del self.listData
            self.listData = tempPtr
        self.length = 0

    def find_item(self, listData, item):
        '''[4]'''
        head = listData
        node = listData
        while node is not None :
            if node.info is item:
                return node
            node = node.next 
            if node.next is head:
                break
        return None
        
    
    def insert_item(self, item):
        '''[5]'''
        newNode = NodeType(item)
        if self.listData is None :
            newNode.next = newNode
            self.listData = newNode
        else :
            node = self.find_item(self.listData, item)
            if node is None :
                node = self.listData
                while node.next is not self.listData :
                    node = node.next
                node.next = newNode
                newNode.next = self.listData
            else:
                nextNode = node.next
                node.next = newNode
                newNode.next = nextNode
        self.length += 1
        return


    def delete_item(self, item):
        '''[6]'''
        node = self.find_item(self.listData, item)
        if node is None :
            return
        back = self.listData
        while back.next is not node :
            back = back.next
        if back is node :
            self.listData = None
        else:
            back.next = node.next
        self.length -= 1
        del node


    def reset_list(self):
        self.currentPos = None

    def get_next_item(self):
        if self.currentPos == None:
            self.currentPos = self.listData
        else:
            self.currentPos = self.currentPos.next
        return self.currentPos.info

    def __str__(self):
        self.reset_list()
        items = []
        for i in range(0, self.length):
            t = self.get_next_item()
            items.append(str(t))
        return " ".join(items)
